[webhook] add webhook module

williamchong opened the pull request:
Current status
- takes a file and json at `file` and `metadata`
- calculates v1 cid of file
- copy input file to config output dir as `${cid}.${ext}`
- parse `metadata` as per [sample json](https://github.com/starlinglab/integrity-schema/blob/main/integrity-backend/input-browsertrix-examples/3e11cc57daf3bad8375935cad4878123acc8d769551ff90f1b1bb0dc597-meta-content.json) which has special property that maybe we should get rid of
  -  contains most metadata at `contentMetadata` Update: now we just use the root level kv pairs
  - has a `private` field to be ignored
- simple aa JSON GET endpoint for `/:cid` and `/:cid/:attr` for verification/debug

makew0rld commented:
Some starting notes before I actually look at the logic. Not all of these are immediately important.

- Using the gin framework seems unnecessary and the standard library [http](https://pkg.go.dev/net/http) could be used instead. I'd prefer to start simple and only add dependencies when needed.
- Instead of having its own `go.mod` and `go.sum`, the `ingest` folder should just be a submodule of `integrity-v2`, and share the dependencies. (We can go through this on a call if it's confusing)
- Similarly, you should move `utils/cid.go` into my `util/` module
- The server should get its configuration from the `config/` module in this repo
- I'm in favour of using the IPFS CLI over calculating it in Go but this will be debated later

williamchong commented:
- used `chi` instead of `gin` since `net/http` doesn't has simple way to route base on method and url params (`/:cid`)
- integrated module as `webhook` as in the name of config
- used existing config for `webhook`, but need clarification on if we need additional config for [special attribute keys](https://github.com/starlinglab/integrity-v2/issues/5)

makew0rld commented:
> used `chi` instead of `gin` since `net/http` (serveMux) doesn't has simple way to route base on method and url params (`/:cid`)

Ok, that makes sense. I've used httprouter in the past and that was good but chi seems pretty equivalent and has no dependencies so that works for me.

makew0rld commented:
Please see the latest version of `aa.go` on main, it should address some of the functions you've created in your PR. You will need to merge main back into your fork here and resolve the conflicts.

I also don't think `CastMapForJSON` should be necessary anymore due to some decoder config I've changed, all decoded maps from CBOR should now be of the type `map[string]any`, not `map[any]any`.

williamchong commented:
> Please see the latest version of `aa.go` on main, it should address some of the functions you've created in your PR. You will need to merge main back into your fork here and resolve the conflicts.
> 
> I also don't think `CastMapForJSON` should be necessary anymore due to some decoder config I've changed, all decoded maps from CBOR should now be of the type `map[string]any`, not `map[any]any`.

Updated and added a function `GetAttestations` for `/c/:cid`

makew0rld reviewed (changes requested):
Thanks for this PR, please see my comments below.

williamchong commented on webhook/webhook.go:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	metadataFields := r.MultipartForm.Value["metadata"]
+	metadataFiles := r.MultipartForm.File["metadata"]
```
the original idea was that folder preprocessor might contain a metadata file alongside the asset
but yes its simpler that we can always make the preprocessor parse it as string field

makew0rld replied:
Yes, even if any kind of preprocessor has the metadata in a file, it should still just send it without a filename.

williamchong replied:
https://github.com/starlinglab/integrity-v2/pull/1/commits/e168ae7760141adb9f537c500e9c72a2d6193c64

williamchong commented on webhook/client.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+func PostFileToWebHook(filePath string, metadata map[string]any) (map[string]any, error) {
+	url, err := urlpkg.Parse(fmt.Sprintf("http://%s/upload", config.GetConfig().Webhook.Host))
```
this one should be `/generic`, but do we want any prefix/suffix for the path? i.e. `/new` `/upload` `/files`, otherwise just `/generic` or `/browsertrix` doesn't seems too meaningful
for project id, it seems to be optional now, since project related data are inside the metadata json. In the future we might need it for matching encryption key, so maybe project should exists as query string

makew0rld replied:
I see. I think it's okay to not prefix these paths since this is all the webhook server does.

> for project id, it seems to be optional now, since project related data are inside the metadata json.

Makes sense. For Browsertrix we will probably want to use it since the Browsertrix webhook request won't tell us the project ID.

williamchong replied:
https://github.com/starlinglab/integrity-v2/pull/1/commits/26f6c4894ffd253830ce0a8f5a0ad7e7af14aab8

williamchong commented on webhook/client.go:
```diff
@@ -0,0 +1,82 @@
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("bad status code in response: %d", resp.StatusCode)
+	}
+	var value map[string]any
+	err = json.NewDecoder(resp.Body).Decode(&value)
+	if err != nil {
+		return nil, err
+	}
+	return value, nil
```
On error it is always `{ error }`, on success currently it just return `{ cid }` (as key) for log usage
Not sure if we want to wrap all success reply payload under some consistent key

makew0rld replied:
I think that response is fine for now. But ideally that would be in a struct, maybe like this:

```go
type Resp struct {
    Cid string `json:"cid,omitempty"`
    Error error `json:"error,omitempty"`
}
```

williamchong replied:
https://github.com/starlinglab/integrity-v2/pull/1/commits/2addb373ac435ca0c2bb883e6a172851c4b0bca3

williamchong commented on aa/aa.go:
```diff
@@ -168,6 +168,22 @@ func GetAttestation(cid, attr string, opts GetAttOpts) (*AttEntry, error) {
 	return &v, nil
 }
 
+func GetAttestations(cid string, opts GetAttOpts) (*map[string]AttEntry, error) {
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/a51274f296c1247bafe7fe4086ca7ff887f3cb7d

williamchong commented on aa/aa.go:
```diff
@@ -168,6 +168,22 @@ func GetAttestation(cid, attr string, opts GetAttOpts) (*AttEntry, error) {
 	return &v, nil
 }
 
+func GetAttestations(cid string, opts GetAttOpts) (*map[string]AttEntry, error) {
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/a51274f296c1247bafe7fe4086ca7ff887f3cb7d

williamchong commented on webhook/client.go:
```diff
@@ -0,0 +1,82 @@
+	if err != nil {
+		return "", nil, err
+	}
+	part, err := mp.CreateFormFile("file", fileInfo.Name())
```
Not meaningful in server side for now, but if we can easily know the filename without extra effort, I just fill it in as best effort/pratice

williamchong replied:
https://github.com/starlinglab/integrity-v2/pull/1/commits/1e8af342a7e278ab44cb75a90727423ae5afaad7

williamchong commented on webhook/client.go:
```diff
@@ -0,0 +1,82 @@
+
+var client = &http.Client{}
+
+func createNewFileForm(filePath string, metadata map[string]any) (string, io.Reader, error) {
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/465ed7145d62d0f81ec3374df650f954dcc6febc

williamchong commented on webhook/file.go:
```diff
@@ -0,0 +1,34 @@
+		outputDirectory = "./output"
+	}
+	if _, err := os.Stat(outputDirectory); os.IsNotExist(err) {
+		err = os.Mkdir(outputDirectory, 0755)
```
Change to throws a error and log
https://github.com/starlinglab/integrity-v2/pull/1/commits/1743654b079bb63f01948d9567f644b43279c505

williamchong commented on webhook/file.go:
```diff
@@ -0,0 +1,34 @@
+	outputDirectory := config.GetConfig().Dirs.Files
+	if outputDirectory == "" {
+		outputDirectory = "./output"
+	}
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/1743654b079bb63f01948d9567f644b43279c505

williamchong commented on webhook/webhook.go:
```diff
@@ -0,0 +1,151 @@
+
+func handleUpload(w http.ResponseWriter, r *http.Request) {
+	// Multipart form
+	err := r.ParseMultipartForm(1024 << 20) // 1024 MB
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/1c6b6a97c07ebc58f784c509fe54143fbebf0275

williamchong commented on aa/aa.go:
```diff
@@ -168,6 +168,22 @@ func GetAttestation(cid, attr string, opts GetAttOpts) (*AttEntry, error) {
+		return nil, err
+	}
+
+	var v map[string]AttEntry
```
https://github.com/starlinglab/integrity-v2/commit/a51274f296c1247bafe7fe4086ca7ff887f3cb7d

williamchong commented on webhook/client.go:
```diff
@@ -0,0 +1,82 @@
+
+func createNewFileForm(filePath string, metadata map[string]any) (string, io.Reader, error) {
+	body := new(bytes.Buffer)
+	mp := multipart.NewWriter(body)
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/4c7f33eae42de2dffcf64ff242c2eb54715f279a

williamchong commented on webhook/file.go:
```diff
@@ -0,0 +1,34 @@
+			return err
+		}
+	}
+	ext := filepath.Ext(originalFileName)
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/781d5762999f45dd304ccb5804a14d58e8825792

williamchong commented on webhook/webhook.go:
```diff
@@ -0,0 +1,151 @@
+	v, err := aa.GetAttestations(cid, aa.GetAttOpts{
+		EncKey:         nil,
+		LeaveEncrypted: false,
+		Format:         "",
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/a51274f296c1247bafe7fe4086ca7ff887f3cb7d

williamchong commented on webhook/webhook.go:
```diff
@@ -0,0 +1,151 @@
+		writeJsonResponse(w, http.StatusBadRequest, map[string]string{"error": "Failed to generate CID for the file."})
+		return
+	}
+	err = CopyOutputToFilePath(&buf, originalFileName, cid)
```
After rewriting multiform parsing, It seems simpler to write whole file to disk first, then calculate cid again

williamchong commented on aa/aa.go:
```diff
@@ -168,6 +168,22 @@ func GetAttestation(cid, attr string, opts GetAttOpts) (*AttEntry, error) {
+	// Ignore format so CBOR is guaranteed
+	opts.Format = ""
+
+	data, err := GetAttestationRaw(cid, "", opts)
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/a51274f296c1247bafe7fe4086ca7ff887f3cb7d

williamchong commented on webhook/webhook.go:
```diff
@@ -0,0 +1,151 @@
+	attributes := ParseJsonToAttributes(jsonMap)
+	err = aa.SetAttestations(cid, false, attributes)
+	if err != nil {
+		writeJsonResponse(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/2cf9ecba4418c8e2ecbd17c96150cf36e62f164e

williamchong commented on webhook/webhook.go:
```diff
@@ -0,0 +1,151 @@
+	r.Get("/ping", handlePing)
+	r.Get("/c/{cid}", handleGetCid)
+	r.Get("/c/{cid}/{attr}", handleGetCidAttribute)
+	r.Post("/upload", handleUpload)
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/47486b47e56ab5531bb88e495fdb63038f510fa5

makew0rld commented on webhook/webhook.go:
```diff
@@ -0,0 +1,200 @@
+		writeJsonResponse(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
+		return
+	}
+	tempFileName := uuid.New()
+	tempFilePath := filepath.Join(outputDirectory, tempFileName.String())
+	fd, err := os.Create(tempFilePath)
```
Using a UUID-named file inside the main files directory as a temporary file will clutter things up and be confusing. It might also be inefficient if the the files directory is an NFS mount or something.

What I've done in some of the CLI tools can be done here:
- Create a temporary file with `os.CreateTemp`
- `defer os.Remove` so that the temporary file is always removed on error
- Use `util.MoveFile` to move it when everything has been written, not `os.Rename`

williamchong replied:
Yes thanks for the hint
https://github.com/starlinglab/integrity-v2/pull/1/commits/f02cefbaab1338f19617c39ec296ae0b3dc27bae

makew0rld commented on webhook/webhook.go:
```diff
@@ -0,0 +1,200 @@
+			}
+		}
+	}
+	err = fd.Close()
+	if err != nil {
+		writeJsonResponse(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
+		return
+	}
+	fd, err = os.Open(tempFilePath)
+	if err != nil {
+		writeJsonResponse(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
+		return
+	}
+	cid := util.CalculateFileCid(fd)
```
Seeking back to the start of the file is probably better than closing and re-opening. Just make sure to `defer fd.Close()` higher up in the code, once it gets opened.

makew0rld replied:
But actually, it should be more efficient to write this file to disk at the same time as the CID is calculated. You can do this in your `io.Copy` line by setting an `io.MultiWriter` as the destination. It would require changing `util.CalculateFileCid` though, to be more like a [Hash](https://pkg.go.dev/hash#Hash).

This isn't strictly necessary but writing and then reading the file could become expensive for multi-gig files. Ideally we would optimize this now.

williamchong replied:
https://github.com/starlinglab/integrity-v2/pull/1/commits/7f9e6fb63fe07495c7a4f787710cea2079328dfd
Please see if this simpler solution with pipe makes sense

makew0rld commented on util/cid.go:
```diff
@@ -0,0 +1,46 @@
+	multicodec "github.com/multiformats/go-multicodec"
+)
+
+func CalculateFileCid(fileReader io.Reader) string {
```
I think it's better for this function to return `(string, error)` so if an error ever happens it can be logged and investigated. And that way callers don't have to check `cid == ""`, they can just check `err != nil` like usual.

makew0rld commented on webhook/webhook.go:
```diff
@@ -0,0 +1,151 @@
+func Run(args []string) error {
+	r := chi.NewRouter()
+	r.Get("/ping", handlePing)
+	r.Get("/c/{cid}", handleGetCid)
+	r.Get("/c/{cid}/{attr}", handleGetCidAttribute)
```
Let's comment these out for now then. I don't like the idea of having this public API available just in case.

makew0rld reviewed (changes requested):
Some smaller fixes now. Thanks!

williamchong commented on webhook/webhook.go:
```diff
@@ -0,0 +1,200 @@
+}
+
+// middleware to validate HMAC JWT token if JWT_SECRET is set
+func jwtAuthMiddleware(next http.Handler) http.Handler {
```
https://github.com/starlinglab/integrity-v2/pull/1/commits/7ac2b0b987ec1b734b7bbfe7b968864edaa03697

makew0rld commented on webhook/webhook.go:
```diff
@@ -0,0 +1,189 @@
+	"github.com/starlinglab/integrity-v2/util"
+)
+
+var jwtSecret = os.Getenv("JWT_SECRET")
+var jwtTokenAuth *jwtauth.JWTAuth
+
+func init() {
+	if jwtSecret != "" {
+		jwtTokenAuth = jwtauth.New("HS256", []byte(jwtSecret), nil)
+	}
+}
```
I would move all of this into `Run`. Also for safety I think `JWT_SECRET` being empty should be an error rather than accepted.

williamchong replied:
maybe we should explicitly set if jwt auth is enabled in webhook server config?

makew0rld commented on webhook/webhook.go:
```diff
@@ -0,0 +1,189 @@
+	// r.Get("/c/{cid}", handleGetCid)
+	// r.Get("/c/{cid}/{attr}", handleGetCidAttribute)
+	r.Route("/generic", func(r chi.Router) {
+		if jwtTokenAuth != nil {
```
Same as above, this should be an error.

makew0rld reviewed (approved):
Looks pretty much done! Just some JWT comments below. If you're happy testing-wise let me know and I will merge.

makew0rld commented:
I will handle any upcoming merge conflicts.

williamchong commented:
@makew0rld Updated and ready to merge

makew0rld commented:
Ok I was unable to push my merge conflict resolution to this PR, so I just merged in into `main` manually in d46dff91e126. You might need to rebase your other PRs off of `main`.
